### PR TITLE
Fix #178: added check for proof of changes

### LINKED ISSUE #178: Oppiabot should ping in the PR thread if "Proof that changes are correct" is left unaddressed
We could have oppiabot ping in the PR thread if "Proof that changes are correct" is left unaddressed.

### PR DESCRIPTION
**Pull request by gp201**
## Explanation

  Fixes #178 
  
  The function checks for the presence of `user-images.githubusercontent.com` since that's where all the images/videos are uploaded.

## Checklist
- [x] I have successfully deployed my own instance of Oppiabot.
  - You can find instructions for doing this [here](https://github.com/oppia/oppiabot/wiki/Deploying-your-own-instance-of-the-oppiabot).
- [x] I have manually tested all the changes made in this PR following the [manual tests matrix](https://github.com/oppia/oppiabot/wiki/Manual-Tests-Matrix).

### CODE REVIEW

**Comment by gp201** [comment]
@vojtechjelinek @jameesjohn PTAL at the approach to solve the issue. Will proceed to write the unit tests after confirmation. Thanks.

I also updates the comment format to display like the following [ref](https://github.com/gp201/Oppiabot-test-repo/pull/11#issuecomment-780464967)

image

**Comment by jameesjohn** [comment]
Nice work @gp201!
Your approach is just correct.

Let me know when you're ready for a full review.

**Comment by gp201** [comment]
@jameesjohn is it okay to change the message format  to the above?

**Comment by gp201** [comment]
All the template problems would now be listed instead of being in one paragraph

**Comment by jameesjohn** [comment]
Yes, that's fine.

**Comment by gp201** [comment]
@jameesjohn PTAL

File: lib/checkPullRequestTemplate.js [hunk 1]
```diff
@@ -165,30 +166,78 @@ const checkEssentialChecklist = (pullRequest) => {
...
+      HEADING_START,
+      proofOfChcangesHeadingIndex + PROOF_TEXT.length
+    );
+
+    let proofOfChangesSection = pullRequest.body.substring(
+      proofOfChcangesHeadingIndex,
+      nextHeadingIndex
+    );
+
+    var proofLinks = proofOfChangesSection.includes(UPLOADED_MEDIA);
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:222]: Hi, I think the problem with this is that sometimes the devs might want to upload pictures in another way than directly via GitHub, for example, a link to the images or videos. So I think it would be enough to check if there is at least something in the section.

**Review by vojtechjelinek — CHANGES_REQUESTED** [review]
Thanks! Left a comment.

File: lib/checkPullRequestTemplate.js (hunk 1 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:222]: @vojtechjelinek could we check for the presence of `http` in that section instead?
Reason:
1. Even an enter or one new character can look like some txt is present.
2. a link will contain a `http` in the beginning of it.

Comment (gp201) [reply lib/checkPullRequestTemplate.js:222]: I tried to check for the presence of other txt by checking if the number of characters of in that section is greater than the number of characters present in the same section in the PR template

File: lib/checkPullRequestTemplate.js [hunk 2]
```diff
@@ -165,30 +166,78 @@ const checkEssentialChecklist = (pullRequest) => {
...
     );
   }
 };
 
+
+/**
+ * This function checks if the pull request's body contains the
+ * checklist and returns an appropriate message if the required
+ * checklist items are not checked.
+ *
```
Comment (jameesjohn) [inline lib/checkPullRequestTemplate.js:191]: I don't think this message is quite right.

File: lib/checkPullRequestTemplate.js (hunk 1 above)
Comment (jameesjohn) [reply lib/checkPullRequestTemplate.js:222]: I think we should check that a valid link is provided.

File: lib/checkPullRequestTemplate.js [hunk 3]
```diff
@@ -203,34 +252,40 @@ const checkTemplate = async (context) => {
...
-    commentBody =
-      'Hi @' +
-      pullRequest.user.login +
-      ', ' +
-      explanationMessage +
-      ' Thanks!';
-  } else if (checklistMessage) {
-    commentBody =
+
+  count = 1;
```
Comment (jameesjohn) [inline lib/checkPullRequestTemplate.js:258]: Where is this variable declared?

File: lib/checkPullRequestTemplate.js [hunk 4]
```diff
@@ -203,34 +252,40 @@ const checkTemplate = async (context) => {
...
+
+  commentBodybasic = commentBody =
```
Comment (jameesjohn) [inline lib/checkPullRequestTemplate.js:260]: Also, this has not been declared.

File: lib/checkPullRequestTemplate.js [hunk 5]
```diff
@@ -203,34 +252,40 @@ const checkTemplate = async (context) => {
...
       'Hi @' +
       pullRequest.user.login +
-      ', ' +
-      checklistMessage.trim() +
-      ' Thanks!';
-  } else {
+      ', can you complete the following:\n';
+
+  if (explanationMessage) {
+    commentBody += count.toString() +
```
Comment (jameesjohn) [inline lib/checkPullRequestTemplate.js:266]: `count.toString()` is not required when concatenating strings.

File: lib/checkPullRequestTemplate.js [hunk 6]
```diff
@@ -203,34 +252,40 @@ const checkTemplate = async (context) => {
...
+    '. ' + proofOfChangesMessage + '\n';
+  }
+
+  if (commentBody === commentBodybasic){
     return;
   }
 
+  commentBody += 'Thanks!';
+
+  commentBody = commentBody.trim();
```
Comment (jameesjohn) [inline lib/checkPullRequestTemplate.js:288]: Can we extract this comment generation into a different function so we can have separate unit tests for it?

**Review by jameesjohn — CHANGES_REQUESTED** [review]
Thanks, @gp201. I left some comments, PTAL

File: lib/checkPullRequestTemplate.js (hunk 1 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:222]: So, do I check for the presence of http? or try a regex to validate the whole url?

Comment (vojtechjelinek) [reply lib/checkPullRequestTemplate.js:222]: Yep checking for the presence of http or https is a smart solution, I'm fine with either checking just for 'http' or for a full URL. 

Also, I just realized that for some changes for example to backend or typing there is no proof of changes needed, so in taht case we should check if the section starts with something like 'No proof of changes needed because' and also edit the PR template to include this info so that people know they always need to fill the section.

Comment (gp201) [reply lib/checkPullRequestTemplate.js:222]: > also edit the PR template to include this info so that people know they always need to fill the section.

```
<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes made in this PR work correctly.
-->
```
The above is already present in the Proof that changes are correct section.

Should we include a point in the PR Pointers section to tell to add `No proof of changes needed because` if no proof are present?

Comment (vojtechjelinek) [reply lib/checkPullRequestTemplate.js:222]: I would add that directly to the section you mention above.

File: lib/checkPullRequestTemplate.js (hunk 2 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:191]: Fixed

File: lib/checkPullRequestTemplate.js (hunk 4 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:260]: Fixed

File: lib/checkPullRequestTemplate.js (hunk 3 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:258]: Fixed

File: lib/checkPullRequestTemplate.js (hunk 6 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js]: DONE. After confirmation, I'll add unit tests

File: lib/checkPullRequestTemplate.js (hunk 5 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js]: DONE

File: lib/checkPullRequestTemplate.js (hunk 1 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js]: Implemented. Will add Tests after confirmation of approach

**Comment by gp201** [comment]
@jameesjohn @vojtechjelinek PTAL. Will add unit tests after the approach is validated. Thanks!

File: lib/checkPullRequestTemplate.js [hunk 7]
```diff
@@ -21,6 +21,11 @@ const { pingAndAssignUsers } = require('./utils');
 const KARMA_LINTER = 'linter/karma';
 const EXPLANATION_TEXT = '[Explain here what your PR does and why]';
 const CHECKLIST_TEXT = '## Essential Checklist';
+const PROOF_TEXT = '## Proof that changes are correct';
+const UPLOADED_MEDIA = 'http';
+const PROOF_UNAVAILABLE_TEXT = 'No proof of changes needed because';
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:26]: Should this be a regex and maybe be a bit less strict. So allow all cases and bigger whitespaces.

File: lib/checkPullRequestTemplate.js [hunk 8]
```diff
@@ -165,69 +169,140 @@ const checkEssentialChecklist = (pullRequest) => {
...
+
+    <!--
+    Add videos/screenshots of the user-facing interface to demonstrate
+    that the changes made in this PR work correctly.
+    -->
+    ![oppiabot-test](https://user-images.githubusercontent.com/30050862/
+      test.gif)
+    (or)
+    https://user-images.githubusercontent.com/30050862/
+      test.mp4
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:210]: What does this mean exactly?

File: lib/checkPullRequestTemplate.js [hunk 9]
```diff
@@ -165,69 +169,140 @@ const checkEssentialChecklist = (pullRequest) => {
...
+      */
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:211]:
```suggestion
  */
```

File: lib/checkPullRequestTemplate.js [hunk 10]
```diff
@@ -165,69 +169,140 @@ const checkEssentialChecklist = (pullRequest) => {
...
+  const proofOfChcangesHeadingIndex = pullRequest.body.indexOf(PROOF_TEXT);
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:212]:
```suggestion
  const proofOfChangesHeadingIndex = pullRequest.body.indexOf(PROOF_TEXT);
```

File: lib/checkPullRequestTemplate.js [hunk 11]
```diff
@@ -165,69 +169,140 @@ const checkEssentialChecklist = (pullRequest) => {
...
+      proofOfChcangesHeadingIndex + PROOF_TEXT.length
+    );
+
+    let proofOfChangesSection = pullRequest.body.substring(
+      proofOfChcangesHeadingIndex,
+      nextHeadingIndex
+    );
+
+    const proofUnavailable = proofOfChangesSection
+      .replace(PROOF_HTML_COMMENT, '').includes(PROOF_UNAVAILABLE_TEXT);
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:226]:
```suggestion
    const proofIsNotPossible = proofOfChangesSection.replace(
      PROOF_HTML_COMMENT, ''
    ).includes(PROOF_UNAVAILABLE_TEXT);
```

File: lib/checkPullRequestTemplate.js [hunk 12]
```diff
@@ -165,69 +169,140 @@ const checkEssentialChecklist = (pullRequest) => {
...
+
+    const proofLinks = proofOfChangesSection.includes(UPLOADED_MEDIA);
+
+    // Check if there are images/videos added in the section.
+    if (!proofLinks && !proofUnavailable){
+      return (
+        'The proof that changes are correct has not been provided, ' +
+        'please make sure to upload a image/video showing that the changes ' +
+        'are correct. ');
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:235]:
```suggestion
        'are correct.'
      );
```

File: lib/checkPullRequestTemplate.js [hunk 13]
```diff
@@ -165,69 +169,140 @@ const checkEssentialChecklist = (pullRequest) => {
...
+    -->
+    ![oppiabot-test](https://user-images.githubusercontent.com/30050862/
+      test.gif)
+    (or)
+    https://user-images.githubusercontent.com/30050862/
+      test.mp4
+      */
+  const proofOfChcangesHeadingIndex = pullRequest.body.indexOf(PROOF_TEXT);
+
+  if (proofOfChcangesHeadingIndex !== -1) {
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:214]:
```suggestion
  if (proofOfChcangesHeadingIndex === -1) {
    return (
      'The body of this PR is missing the proof that changes ' +
      'are correct section, please update it to include the section.'
    );
  }
```

Comment (vojtechjelinek) [reply lib/checkPullRequestTemplate.js:214]: And you can remove the code from below.

**Review by vojtechjelinek — CHANGES_REQUESTED** [review]
Thanks! Left a few comments.

File: lib/checkPullRequestTemplate.js (hunk 13 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:214]: Wouldn't be better to follow the same format as the other PR template checks, where the above comment would be returned only in the else section?

File: lib/checkPullRequestTemplate.js (hunk 8 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:210]: These are the links that will be generated if an image/video is directly uploaded to the PR template

File: lib/checkPullRequestTemplate.js (hunk 7 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:26]: Fixed.
The following cases are now covered:
image

**Comment by gp201** [comment]
@vojtechjelinek @jameesjohn PTAL

File: lib/checkPullRequestTemplate.js (hunk 7 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js]: it is case-insensitive and allows any number of white spaces in between

File: lib/checkPullRequestTemplate.js (hunk 13 above)
Comment (vojtechjelinek) [reply lib/checkPullRequestTemplate.js]: I think that instead of having a big nested block it is better to keep the nested sections smaller.

File: lib/checkPullRequestTemplate.js (hunk 8 above)
Comment (vojtechjelinek) [reply lib/checkPullRequestTemplate.js:211]: I get that but I don't understand the formatting here.

File: lib/checkPullRequestTemplate.js [hunk 14]
```diff
@@ -165,69 +170,144 @@ const checkEssentialChecklist = (pullRequest) => {
...
+
+    <!--
+    Add videos/screenshots of the user-facing interface to demonstrate
+    that the changes made in this PR work correctly.
+    -->
+    ![oppiabot-test](https://user-images.githubusercontent.com/30050862/
+      test.gif)
+    (or)
+    https://user-images.githubusercontent.com/30050862/
+      test.mp4
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:211]: Let's rephrase it to something like
```
A valid 'Proof that changes are correct' section would either:
  1. Contain a link to some image or video that contains the proof.
  2. Contain a text saying 'No proof of changes needed because …'
```

File: spec/checkPullRequestTemplateSpec.js [hunk 15]
```diff
@@ -167,10 +218,43 @@ describe('Pull Request Template', () => {
...
+
+  const incompleteChecklistSection1 =
+    'The karma and linter checklist has not been checked, ' +
+    'please make sure to run the frontend tests and lint tests before ' +
+    'pushing. ';
+
+  const incompleteChecklistSection2 =
+    'The allow edits from maintainers checklist needs to ' +
+    'be ticked so that maintainers can rerun failed tests. ' +
+    'Endeavour to add this by ticking on the check box. ';
```
Comment (vojtechjelinek) [inline spec/checkPullRequestTemplateSpec.js:243]: Why not merge these constants together?

File: lib/checkPullRequestTemplate.js [hunk 16]
```diff
@@ -165,69 +170,144 @@ const checkEssentialChecklist = (pullRequest) => {
...
+  }
+
+  if (proofOfChangesMessage) {
+    commentBody += count +
+    '. ' + proofOfChangesMessage + '\n';
+  }
+
+  if (commentBody === commentBodybasic){
+    return;
+  }
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:289]:
```suggestion
  // If there is no reason to produce the message do not return anything.
  if (count === 1) {
    return;
  }
```

File: lib/checkPullRequestTemplate.js [hunk 17]
```diff
@@ -165,69 +170,144 @@ const checkEssentialChecklist = (pullRequest) => {
...
+
+  commentBody += 'Thanks!';
+
+  commentBody = commentBody.trim();
+  return commentBody;
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:294]:
```suggestion
  return commentBody.trim();
```

File: lib/checkPullRequestTemplate.js [hunk 18]
```diff
@@ -165,69 +170,144 @@ const checkEssentialChecklist = (pullRequest) => {
...
  * This function checks that a PR follows the appropriate template and
  * comments on the PR if key parts of the template is missing.
  *
  * @param {import('probot').Context} context
  */
+const generateComment = (pullRequest) => {
+  const explanationMessage = checkPRExplanation(pullRequest);
+  const checklistMessage = checkEssentialChecklist(pullRequest);
+  const proofOfChangesMessage = checkProofThatChangesAreCorrect(pullRequest);
+  let commentBody = '';
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:261]:
```suggestion
  let commentBody = (
    'Hi @' + pullRequest.user.login + ', can you complete the following:\n');
```

File: lib/checkPullRequestTemplate.js [hunk 19]
```diff
@@ -165,69 +170,144 @@ const checkEssentialChecklist = (pullRequest) => {
...
+
+  var count = 1;
+
+  var commentBodybasic = commentBody =
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:265]:
```suggestion
  let count = 1;

  let commentBodyBasic = commentBody;
```

File: lib/checkPullRequestTemplate.js [hunk 20]
```diff
@@ -165,69 +170,144 @@ const checkEssentialChecklist = (pullRequest) => {
...
+      'Hi @' +
+      pullRequest.user.login +
+      ', can you complete the following:\n';
+
+  if (explanationMessage) {
+    commentBody += count +
+    '. ' + explanationMessage + '\n';
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:272]:
```suggestion
    commentBody += count + '. ' + explanationMessage + '\n';
```

File: lib/checkPullRequestTemplate.js [hunk 21]
```diff
@@ -165,69 +170,144 @@ const checkEssentialChecklist = (pullRequest) => {
...
+    count += 1;
+  }
+
+  if (checklistMessage) {
+    commentBody += count +
+    '. ' + checklistMessage + '\n';
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:278]:
```suggestion
    commentBody += count + '. ' + checklistMessage + '\n';
```

File: lib/checkPullRequestTemplate.js [hunk 22]
```diff
@@ -165,69 +170,144 @@ const checkEssentialChecklist = (pullRequest) => {
...
+    count += 1;
+  }
+
+  if (proofOfChangesMessage) {
+    commentBody += count +
+    '. ' + proofOfChangesMessage + '\n';
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:284]:
```suggestion
    commentBody += count + '. ' + proofOfChangesMessage + '\n';
    count += 1;
```

File: lib/checkPullRequestTemplate.js [hunk 23]
```diff
@@ -165,69 +170,144 @@ const checkEssentialChecklist = (pullRequest) => {
...
-      ' Thanks!';
-  } else if (checklistMessage) {
-    commentBody =
-      'Hi @' +
-      pullRequest.user.login +
-      ', ' +
-      checklistMessage.trim() +
-      ' Thanks!';
-  } else {
+  if (!commentBody){
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:310]:
```suggestion
  if (!commentBody) {
```

File: lib/checkPullRequestTemplate.js [hunk 24]
```diff
@@ -165,69 +170,144 @@ const checkEssentialChecklist = (pullRequest) => {
...
+      PROOF_HTML_COMMENT, ''
+    ).match(PROOF_UNAVAILABLE_TEXT));
+    const proofLinks = proofOfChangesSection.includes(UPLOADED_MEDIA);
+
+    // Check if there are images/videos added in the section.
+    if (!proofLinks && !proofIsNotPossible){
+      return (
+        'The proof that changes are correct has not been provided, ' +
+        'please make sure to upload a image/video showing that the changes ' +
+        'are correct.'
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:239]:
```suggestion
        'The proof that changes are correct has not been provided, ' +
        'please make sure to upload a image/video showing that the changes ' +
        'are correct. Or include a sentence saying "No proof of changes' +
        'needed because" and the reason why proof of changes cannot be provided.'
```

File: lib/checkPullRequestTemplate.js [hunk 25]
```diff
@@ -165,69 +170,144 @@ const checkEssentialChecklist = (pullRequest) => {
...
+    const proofIsNotPossible = proofOfChangesSection.replace(
+      PROOF_HTML_COMMENT, ''
+    ).match(PROOF_UNAVAILABLE_TEXT);
+    console.log(proofOfChangesSection.replace(
+      PROOF_HTML_COMMENT, ''
+    ).match(PROOF_UNAVAILABLE_TEXT));
+    const proofLinks = proofOfChangesSection.includes(UPLOADED_MEDIA);
+
+    // Check if there are images/videos added in the section.
+    if (!proofLinks && !proofIsNotPossible){
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:235]:
```suggestion
    if (!proofLinks && !proofIsNotPossible) {
```

**Review by vojtechjelinek — CHANGES_REQUESTED** [review]
Thanks! Left a few other comments.

File: spec/checkPullRequestTemplateSpec.js (hunk 15 above)
Comment (gp201) [reply spec/checkPullRequestTemplateSpec.js:243]: Into an array?

File: lib/checkPullRequestTemplate.js (hunk 14 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:211]: DONE

File: spec/checkPullRequestTemplateSpec.js (hunk 15 above)
Comment (gp201) [reply spec/checkPullRequestTemplateSpec.js:243]: DONE

File: lib/checkPullRequestTemplate.js (hunk 16 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:289]: DONE

File: lib/checkPullRequestTemplate.js (hunk 17 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:294]: DONE

File: lib/checkPullRequestTemplate.js (hunk 18 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:261]: DONE

File: lib/checkPullRequestTemplate.js (hunk 19 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:265]: DONE

File: lib/checkPullRequestTemplate.js (hunk 20 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:272]: DONE

File: lib/checkPullRequestTemplate.js (hunk 21 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:278]: DONE

File: lib/checkPullRequestTemplate.js (hunk 22 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:284]: DONE

File: lib/checkPullRequestTemplate.js (hunk 23 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:310]: DONE

File: lib/checkPullRequestTemplate.js (hunk 24 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:239]: DONE

File: lib/checkPullRequestTemplate.js (hunk 25 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:235]: DONE

File: lib/checkPullRequestTemplate.js (hunk 13 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js]: DONE

**Comment by gp201** [comment]
@jameesjohn @vojtechjelinek PTAL. Will add unit tests after the approach is validated. Thanks!

File: lib/checkPullRequestTemplate.js [hunk 26]
```diff
@@ -165,69 +170,146 @@ const checkEssentialChecklist = (pullRequest) => {
...
  *
  * @param {import('probot').Context} context
  */
+const generateComment = (pullRequest) => {
+  const explanationMessage = checkPRExplanation(pullRequest);
+  const checklistMessage = checkEssentialChecklist(pullRequest);
+  const proofOfChangesMessage = checkProofThatChangesAreCorrect(pullRequest);
+
+  let count = 1;
+
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:270]: nit: unneeded empty line

File: lib/checkPullRequestTemplate.js [hunk 27]
```diff
@@ -165,69 +170,146 @@ const checkEssentialChecklist = (pullRequest) => {
...
+    count += 1;
+  }
+
+  // If there is no reason to produce the message do not return anything.
+  if (count === 1) {
+    return;
+  }
+
+  commentBody += 'Thanks!';
+
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:295]: nit: unneeded empty line

File: lib/checkPullRequestTemplate.js [hunk 28]
```diff
@@ -165,69 +170,146 @@ const checkEssentialChecklist = (pullRequest) => {
...
 
+
+/**
+ * This function checks if the pull request's body contains the
+ * proof of changes and returns an appropriate message if the required
+ * proofs are not provided.
+ *
+ * @param {import('probot').Octokit.PullsGetResponse} pullRequest
+ */
+const checkProofThatChangesAreCorrect = (pullRequest) => {
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:198]:
```suggestion
const checkProofOfChanges = (pullRequest) => {
```

File: lib/checkPullRequestTemplate.js [hunk 29]
```diff
@@ -165,69 +170,146 @@ const checkEssentialChecklist = (pullRequest) => {
...
+    proofOfChangesHeadingIndex,
+    nextHeadingIndex
+  );
+
+  const proofIsNotPossible = proofOfChangesSection.replace(
+    PROOF_HTML_COMMENT, ''
+  ).match(PROOF_UNAVAILABLE_TEXT);
+  console.log(proofOfChangesSection.replace(
+    PROOF_HTML_COMMENT, ''
+  ).match(PROOF_UNAVAILABLE_TEXT));
```
Comment (vojtechjelinek) [inline lib/checkPullRequestTemplate.js:242]: Remove this log.

**Review by vojtechjelinek — CHANGES_REQUESTED** [review]
Thanks! Only a last few comments.

File: lib/checkPullRequestTemplate.js (hunk 26 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:270]: DONE

File: lib/checkPullRequestTemplate.js (hunk 27 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:295]: DONE

File: lib/checkPullRequestTemplate.js (hunk 28 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:198]: DONE

File: lib/checkPullRequestTemplate.js (hunk 29 above)
Comment (gp201) [reply lib/checkPullRequestTemplate.js:242]: DONE

**Comment by gp201** [comment]
@jameesjohn @vojtechjelinek PTAL

**Comment by gp201** [comment]
@jameesjohn You had mentioned [here](https://github.com/oppia/oppiabot/pull/202#discussion_r579158847) that we're going to have separate unit tests for the function. We can't test an anonymous function, right?

**Review by vojtechjelinek — APPROVED** [review]
LGTM!

**Review by jameesjohn — CHANGES_REQUESTED** [review]

**Comment by jameesjohn** [comment]
>  We can't test an anonymous function, right?

You should be able to test the function after you export it.

File: spec/checkPullRequestTemplateSpec.js [hunk 30]
```diff
@@ -494,12 +653,58 @@ describe('Pull Request Template', () => {
...
+        'Hi @' +
+        payloadData.payload.pull_request.user.login +
+        ', can you complete the following:' + '\n' +
+        '1. ' + incompleteChecklistSection.karma_linter +
+        '\n' + '2. ' + incompleteProofOfChangesSection + '\nThanks!',
+      });
+    });
+  });
+
+  describe('testing generateComment function with bodyWithInvalidTemplate',
```
Comment (gp201) [inline spec/checkPullRequestTemplateSpec.js:665]: @jameesjohn PTAL if these tests are correct. I'll do the rest of the tests after confirmation since the rest of the tests will follow the same format.

File: lib/checkPullRequestTemplate.js [hunk 31]
```diff
@@ -21,6 +21,12 @@ const { pingAndAssignUsers } = require('./utils');
 const KARMA_LINTER = 'linter/karma';
 const EXPLANATION_TEXT = '[Explain here what your PR does and why]';
 const CHECKLIST_TEXT = '## Essential Checklist';
+const PROOF_TEXT = '## Proof that changes are correct';
+const UPLOADED_MEDIA = 'http';
```
[the remark this answers is not in the document]
Comment (gp201) [reply lib/checkPullRequestTemplate.js:25]: DONE

File: lib/checkPullRequestTemplate.js [hunk 32]
```diff
@@ -21,6 +21,12 @@ const { pingAndAssignUsers } = require('./utils');
...
+const PROOF_UNAVAILABLE_TEXT =
```
[the remark this answers is not in the document]
Comment (gp201) [reply lib/checkPullRequestTemplate.js:26]: DONE

File: spec/checkPullRequestTemplateSpec.js (hunk 30 above)
Comment (jameesjohn) [reply spec/checkPullRequestTemplateSpec.js:665]: Yes, it looks good.

**Review by jameesjohn — APPROVED** [review]
LGTM
